File: runtime-c/src/smtlib.cpp
#include "smtlib.h"

#include <optional>
#include <sstream>

namespace evc {

std::string Sort::smt() const {
    switch (tag) {
        case Tag::Int:  return "Int";
        case Tag::Bool: return "Bool";
        case Tag::Real: return "Real";
        case Tag::Str:  return "String";
        case Tag::Enum: return enum_name;
    }
    return "Int";
}

namespace {

// Map an Evident scalar type name to an SMT sort. Nat/Pos are Int with an added
// bound emitted by the caller. Enum sorts handled separately by the emitter.
std::optional<Sort> scalar_sort(const std::string &type_name) {
    if (type_name == "Int" || type_name == "Nat" || type_name == "Pos") return Sort{Sort::Tag::Int, {}};
    if (type_name == "Bool") return Sort{Sort::Tag::Bool, {}};
    if (type_name == "Real") return Sort{Sort::Tag::Real, {}};
    if (type_name == "String") return Sort{Sort::Tag::Str, {}};
    return std::nullopt;
}

// SMT-LIB int literal — negatives wrap in (- n). Magnitude computed via unsigned
// to dodge INT64_MIN overflow on negate.
std::string int_lit_safe(int64_t i) {
    if (i >= 0) return std::to_string(i);
    unsigned long long mag = (unsigned long long)(-(i + 1)) + 1ULL;
    return "(- " + std::to_string(mag) + ")";
}

std::string real_lit(double r) {
    std::ostringstream os;
    double mag = r < 0 ? -r : r;
    os << mag;
    std::string s = os.str();
    if (s.find('.') == std::string::npos && s.find('e') == std::string::npos &&
        s.find('E') == std::string::npos)
        s += ".0";
    if (r < 0) return "(- " + s + ")";
    return s;
}

std::string str_lit(const std::string &s) {
    std::string out = "\"";
    for (char c : s) {
        if (c == '"') out += '"';  // SMT-LIB doubles internal quotes
        out += c;
    }
    out += "\"";
    return out;
}

// ---------------------------------------------------------------------------
// Emitter: SchemaDecl -> SMT-LIB. Stateful so enum datatypes + quantifier
// unrolling (M4) can extend it without re-threading parameters.
// ---------------------------------------------------------------------------
class Emitter {
public:
    explicit Emitter(const Program &prog) : prog_(prog) { index_enums(); }

    EmitResult emit(const SchemaDecl &schema);

private:
    const Program &prog_;
    std::unordered_map<std::string, Sort> env_;
    std::vector<std::pair<std::string, Sort>> declared_;
    std::string out_;

    // enum registries (built in M4; scalar subset leaves them empty)
    std::unordered_map<std::string, const EnumDecl *> enum_by_name_;
    // variant name -> (enum name, field count)
    std::unordered_map<std::string, std::pair<std::string, size_t>> variant_;

    void index_enums();
    void emit_datatypes();

    std::optional<Sort> sort_of(const Expr &e);
    std::string expr(const Expr &e);
    std::string binary(BinOp op, const Expr &a, const Expr &b);
    std::string in_expr(const Expr &lhs, const Expr &rhs);

    [[noreturn]] void fail(const std::string &m) { throw SmtError(m); }
};

void Emitter::index_enums() {
    for (const auto &e : prog_.enums) {
        enum_by_name_[e.name] = &e;
        for (const auto &v : e.variants)
            variant_[v.name] = {e.name, v.fields.size()};
    }
}

void Emitter::emit_datatypes() {
    // M4: emit declare-datatypes for enums. Scalar subset: nothing.
    // (Implemented when enums are wired; left as a no-op preamble for now.)
}

EmitResult Emitter::emit(const SchemaDecl &schema) {
    emit_datatypes();

    // Pass 1: declarations.
    for (const auto &item : schema.body) {
        if (item.kind != BodyItem::Kind::Membership) continue;
        const std::string &name = item.name;
        const std::string &tn = item.type_name;

        std::optional<Sort> sort = scalar_sort(tn);
        if (!sort) {
            // enum type?
            auto it = enum_by_name_.find(tn);
            if (it != enum_by_name_.end()) sort = Sort{Sort::Tag::Enum, tn};
        }
        if (!sort) fail("unsupported type `" + tn + "` for `" + name + "`");

        if (env_.count(name)) continue;  // first decl wins
        env_[name] = *sort;
        declared_.push_back({name, *sort});
        out_ += "(declare-const " + name + " " + sort->smt() + ")\n";
        if (tn == "Nat") out_ += "(assert (>= " + name + " 0))\n";
        else if (tn == "Pos") out_ += "(assert (> " + name + " 0))\n";

        if (item.pins.kind != Pins::Kind::None)
            fail("pins on scalar `" + name + "` not supported");
    }

    // Pass 2: constraints.
    for (const auto &item : schema.body) {
        switch (item.kind) {
            case BodyItem::Kind::Membership: break;
            case BodyItem::Kind::Constraint:
                out_ += "(assert " + expr(*item.expr) + ")\n";
                break;
            case BodyItem::Kind::Passthrough: fail("passthrough `.." + item.name + "` not supported");
            case BodyItem::Kind::Subclaim: fail("subclaim not supported");
            case BodyItem::Kind::ClaimCall: fail("claim call `" + item.name + "` not supported");
            case BodyItem::Kind::HaltsWithin: fail("halts_within not supported");
        }
    }

    return EmitResult{out_, declared_};
}

std::optional<Sort> Emitter::sort_of(const Expr &e) {
    using K = Expr::Kind;
    switch (e.kind) {
        case K::Int:  return Sort{Sort::Tag::Int, {}};
        case K::Real: return Sort{Sort::Tag::Real, {}};
        case K::Bool: return Sort{Sort::Tag::Bool, {}};
        case K::Str:  return Sort{Sort::Tag::Str, {}};
        case K::Ident: {
            auto it = env_.find(e.str);
            if (it != env_.end()) return it->second;
            auto v = variant_.find(e.str);
            if (v != variant_.end()) return Sort{Sort::Tag::Enum, v->second.first};
            return std::nullopt;
        }
        case K::Not: return Sort{Sort::Tag::Bool, {}};
        case K::In: case K::Matches: return Sort{Sort::Tag::Bool, {}};
        case K::Binary:
            switch (e.op) {
                case BinOp::Eq: case BinOp::Neq: case BinOp::Lt: case BinOp::Le:
                case BinOp::Gt: case BinOp::Ge: case BinOp::And: case BinOp::Or:
                case BinOp::Implies:
                    return Sort{Sort::Tag::Bool, {}};
                case BinOp::Concat: return Sort{Sort::Tag::Str, {}};
                case BinOp::Add: case BinOp::Sub: case BinOp::Mul: case BinOp::Div: {
                    auto a = sort_of(*e.children[0]);
                    auto b = sort_of(*e.children[1]);
                    if ((a && a->tag == Sort::Tag::Real) || (b && b->tag == Sort::Tag::Real))
                        return Sort{Sort::Tag::Real, {}};
                    if ((a && a->tag == Sort::Tag::Int) || (b && b->tag == Sort::Tag::Int))
                        return Sort{Sort::Tag::Int, {}};
                    return std::nullopt;
                }
            }
            return std::nullopt;
        case K::Ternary: {
            auto t = sort_of(*e.children[1]);
            return t ? t : sort_of(*e.children[2]);
        }
        case K::Call: {
            auto v = variant_.find(e.str);
            if (v != variant_.end()) return Sort{Sort::Tag::Enum, v->second.first};
            return std::nullopt;
        }
        default: return std::nullopt;
    }
}

std::string Emitter::expr(const Expr &e) {
    using K = Expr::Kind;
    switch (e.kind) {
        case K::Ident: {
            if (env_.count(e.str)) return e.str;
            auto v = variant_.find(e.str);
            if (v != variant_.end() && v->second.second == 0) return e.str;  // nullary ctor
            fail("undeclared identifier `" + e.str + "` (out of scalar subset)");
        }
        case K::Int:  return int_lit_safe(e.ival);
        case K::Real: return real_lit(e.rval);
        case K::Bool: return e.bval ? "true" : "false";
        case K::Str:  return str_lit(e.str);
        case K::Not:  return "(not " + expr(*e.children[0]) + ")";
        case K::Binary: return binary(e.op, *e.children[0], *e.children[1]);
        case K::Ternary:
            return "(ite " + expr(*e.children[0]) + " " + expr(*e.children[1]) +
                   " " + expr(*e.children[2]) + ")";
        case K::In: return in_expr(*e.children[0], *e.children[1]);

        // Out of scalar subset (some wired in M4).
        case K::SetLit:  fail("set literal (not as ∈ RHS) unsupported");
        case K::SeqLit:  fail("sequence literal unsupported");
        case K::Range:   fail("bare range unsupported (only as ∈ RHS)");
        case K::Tuple:   fail("tuple unsupported");
        case K::Forall:  fail("∀ quantifier unsupported (quantifier-free subset)");
        case K::Exists:  fail("∃ quantifier unsupported (quantifier-free subset)");
        case K::Call:    fail("call `" + e.str + "` unsupported");
        case K::Card:    fail("cardinality `#` unsupported");
        case K::Index:   fail("indexing `[]` unsupported");
        case K::Field:   fail("field access unsupported (records out of subset)");
        case K::Match:   fail("match unsupported");
        case K::Matches: fail("matches recognizer unsupported");
    }
    fail("unreachable expr kind");
}

std::string Emitter::binary(BinOp op, const Expr &a, const Expr &b) {
    if (op == BinOp::Neq)
        return "(not (= " + expr(a) + " " + expr(b) + "))";
    std::string sym;
    switch (op) {
        case BinOp::Eq: sym = "="; break;
        case BinOp::Lt: sym = "<"; break;
        case BinOp::Le: sym = "<="; break;
        case BinOp::Gt: sym = ">"; break;
        case BinOp::Ge: sym = ">="; break;
        case BinOp::And: sym = "and"; break;
        case BinOp::Or: sym = "or"; break;
        case BinOp::Implies: sym = "=>"; break;
        case BinOp::Add: sym = "+"; break;
        case BinOp::Sub: sym = "-"; break;
        case BinOp::Mul: sym = "*"; break;
        case BinOp::Concat: sym = "str.++"; break;
        case BinOp::Div: {
            // Int division -> div; Real -> /. Infer from operand sorts.
            auto sa = sort_of(a);
            auto sb = sort_of(b);
            bool real = (sa && sa->tag == Sort::Tag::Real) || (sb && sb->tag == Sort::Tag::Real);
            sym = real ? "/" : "div";
            break;
        }
        case BinOp::Neq: break;  // handled above
    }
    return "(" + sym + " " + expr(a) + " " + expr(b) + ")";
}

std::string Emitter::in_expr(const Expr &lhs, const Expr &rhs) {
    std::string l = expr(lhs);
    if (rhs.kind == Expr::Kind::Range) {
        std::string lo = expr(*rhs.children[0]);
        std::string hi = expr(*rhs.children[1]);
        return "(and (>= " + l + " " + lo + ") (<= " + l + " " + hi + "))";
    }
    if (rhs.kind == Expr::Kind::SetLit) {
        if (rhs.children.empty()) return "false";
        std::vector<std::string> parts;
        for (const auto &el : rhs.children)
            parts.push_back("(= " + l + " " + expr(*el) + ")");
        if (parts.size() == 1) return parts[0];
        std::string s = "(or";
        for (auto &p : parts) s += " " + p;
        s += ")";
        return s;
    }
    fail("∈ RHS must be a set literal or range in the scalar subset");
}

}  // namespace

EmitResult emit_schema(const SchemaDecl &schema, const Program &prog) {
    Emitter em(prog);
    return em.emit(schema);
}

std::string schema_to_smtlib(const SchemaDecl &schema, const Program &prog) {
    return emit_schema(schema, prog).text;
}

}  // namespace evc

File: runtime/src/translate/smtlib.rs
//! PROTOTYPE — Evident claim → SMT-LIB text → Z3 (parse) → solve.
//!
//! This is the first concrete slice of the north star in
//! `docs/design/smtlib-as-compile-target.md`: instead of building the Z3 AST
//! through the C API (what the rest of `translate/` does), this module emits
//! **SMT-LIB text** for a quantifier-free scalar subset, hands it to Z3's own
//! parser (`Z3_solver_from_string`), and solves. The round-trip proof lives in
//! `runtime/tests/smtlib_roundtrip.rs`: for a corpus of simple claims it asserts
//! that this path produces the SAME sat/unsat (and, where the model is forced,
//! the same model) as the default `EvidentRuntime::query` path.
//!
//! It is **additive and gated**: nothing on the default translate/query path
//! calls into here. The only callers are the round-trip test (the "dedicated
//! test entry") and anyone who opts in via [`is_enabled`] (`EVIDENT_SMTLIB=1`).
//!
//! Subset that transpiles today (see the findings doc for the full table):
//!   * scalar sorts `Int` / `Nat` / `Pos` / `Bool` / `Real` / `String`
//!   * arithmetic `+ - * /`, comparisons `= ≠ < ≤ > ≥`, `∧ ∨ ¬ ⇒`
//!   * set/range membership as a constraint (`x ∈ {1,2,3}`, `x ∈ {lo..hi}`)
//!   * ternary `(c ? a : b)` → `ite`
//! Anything else (Seq/Set values, enums, records, quantifiers, match, FSM
//! machinery, claim composition) returns [`SmtLibError`] — the boundary is
//! reported honestly, never silently mistranslated.

use std::collections::HashMap;
use std::fmt::Write as _;

use z3::{Config, Context, SatResult, Solver};

use crate::core::ast::{BinOp, BodyItem, Expr, Pins, SchemaDecl};
use crate::core::Value;

/// Out-of-subset / malformed-input signal. The prototype reports exactly what it
/// could not transpile rather than emitting wrong SMT-LIB.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SmtLibError(pub String);

impl std::fmt::Display for SmtLibError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "smtlib: {}", self.0)
    }
}
impl std::error::Error for SmtLibError {}

fn err<T>(msg: impl Into<String>) -> Result<T, SmtLibError> {
    Err(SmtLibError(msg.into()))
}

/// The scalar SMT sorts the prototype handles.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Sort {
    Int,
    Bool,
    Real,
    Str,
}

impl Sort {
    /// SMT-LIB sort name.
    fn smt(self) -> &'static str {
        match self {
            Sort::Int => "Int",
            Sort::Bool => "Bool",
            Sort::Real => "Real",
            Sort::Str => "String",
        }
    }
}

/// Map an Evident scalar type name to an SMT sort. `Nat`/`Pos` are `Int` with an
/// extra non-negativity / positivity bound emitted by the caller.
fn scalar_sort(type_name: &str) -> Option<Sort> {
    match type_name {
        "Int" | "Nat" | "Pos" => Some(Sort::Int),
        "Bool" => Some(Sort::Bool),
        "Real" => Some(Sort::Real),
        "String" => Some(Sort::Str),
        _ => None,
    }
}

/// Variable name → sort, built from the schema's `Membership` body items.
type Env = HashMap<String, Sort>;

/// Result of running a claim through the SMT-LIB path.
#[derive(Debug, Clone)]
pub struct SmtSolveResult {
    pub satisfied: bool,
    /// Extracted model for declared scalar consts (only populated when satisfied).
    pub bindings: HashMap<String, Value>,
    /// The SMT-LIB text that was handed to Z3 — useful for tests / debugging.
    pub smtlib: String,
}

/// Opt-in flag for anyone wiring the prototype into a non-test path. The default
/// translate/query path never consults this; it exists so the env var named in
/// the design doc has a single source of truth.
pub fn is_enabled() -> bool {
    std::env::var("EVIDENT_SMTLIB").map(|s| s != "0" && !s.is_empty()).unwrap_or(false)
}

// ---------------------------------------------------------------------------
// Emit: SchemaDecl → SMT-LIB text
// ---------------------------------------------------------------------------

/// Emit SMT-LIB text (`declare-const` + `assert` lines, no `check-sat`) for a
/// claim's free-query semantics: every declared scalar is a fresh const, every
/// constraint is asserted. Returns the declared `(name, sort)` list alongside.
fn emit(schema: &SchemaDecl) -> Result<(String, Vec<(String, Sort)>), SmtLibError> {
    let mut env: Env = HashMap::new();
    let mut declared: Vec<(String, Sort)> = Vec::new();
    let mut out = String::new();

    // Pass 1: declarations (so forward references in constraints resolve).
    for item in &schema.body {
        if let BodyItem::Membership { name, type_name, pins } = item {
            let sort = scalar_sort(type_name)
                .ok_or_else(|| SmtLibError(format!("unsupported type `{type_name}` for `{name}`")))?;
            if env.contains_key(name) {
                continue; // re-declaration; first wins (matches declare.rs idempotence)
            }
            env.insert(name.clone(), sort);
            declared.push((name.clone(), sort));
            let _ = writeln!(out, "(declare-const {name} {})", sort.smt());
            // Nat/Pos lower bound.
            match type_name.as_str() {
                "Nat" => { let _ = writeln!(out, "(assert (>= {name} 0))"); }
                "Pos" => { let _ = writeln!(out, "(assert (> {name} 0))"); }
                _ => {}
            }
            if !matches!(pins, Pins::None) {
                return err(format!("pins on scalar `{name}` not supported in prototype"));
            }
        }
    }

    // Pass 2: constraints.
    for item in &schema.body {
        match item {
            BodyItem::Membership { .. } => {} // handled above
            BodyItem::Constraint(e) => {
                let s = expr(e, &env)?;
                let _ = writeln!(out, "(assert {s})");
            }
            BodyItem::Passthrough(n) => return err(format!("passthrough `..{n}` not supported")),
            BodyItem::SubclaimDecl(_) => return err("subclaim not supported"),
            BodyItem::ClaimCall { name, .. } => return err(format!("claim call `{name}` not supported")),
            BodyItem::HaltsWithin { .. } => return err("halts_within not supported"),
        }
    }

    Ok((out, declared))
}

/// Public entry: just the SMT-LIB text (for inspection / tests).
pub fn schema_to_smtlib(schema: &SchemaDecl) -> Result<String, SmtLibError> {
    emit(schema).map(|(text, _)| text)
}

// ---------------------------------------------------------------------------
// Expr → SMT-LIB s-expression
// ---------------------------------------------------------------------------

/// Best-effort sort inference for an expression, used to pick `div` vs `/` and
/// to validate branch sorts. `None` when the prototype can't determine it.
fn sort_of(e: &Expr, env: &Env) -> Option<Sort> {
    match e {
        Expr::Int(_) => Some(Sort::Int),
        Expr::Real(_) => Some(Sort::Real),
        Expr::Bool(_) => Some(Sort::Bool),
        Expr::Str(_) => Some(Sort::Str),
        Expr::Identifier(n) => env.get(n).copied(),
        Expr::Not(_) => Some(Sort::Bool),
        Expr::Binary(op, a, b) => match op {
            BinOp::Eq | BinOp::Neq | BinOp::Lt | BinOp::Le | BinOp::Gt | BinOp::Ge
            | BinOp::And | BinOp::Or | BinOp::Implies => Some(Sort::Bool),
            BinOp::Concat => Some(Sort::Str),
            BinOp::Add | BinOp::Sub | BinOp::Mul | BinOp::Div => {
                // Arithmetic sort = Real if either operand is Real, else Int.
                match (sort_of(a, env), sort_of(b, env)) {
                    (Some(Sort::Real), _) | (_, Some(Sort::Real)) => Some(Sort::Real),
                    (Some(Sort::Int), _) | (_, Some(Sort::Int)) => Some(Sort::Int),
                    _ => None,
                }
            }
        },
        Expr::Ternary(_, t, f) => sort_of(t, env).or_else(|| sort_of(f, env)),
        Expr::InExpr(..) => Some(Sort::Bool),
        _ => None,
    }
}

/// Translate one expression to an SMT-LIB s-expression string.
fn expr(e: &Expr, env: &Env) -> Result<String, SmtLibError> {
    match e {
        Expr::Identifier(n) => {
            if env.contains_key(n) {
                Ok(n.clone())
            } else {
                // An undeclared bare identifier is almost always a record field /
                // enum ctor / out-of-subset name. Report rather than emit garbage.
                err(format!("undeclared identifier `{n}` (out of scalar subset)"))
            }
        }
        Expr::Int(i) => Ok(int_lit(*i)),
        Expr::Real(r) => Ok(real_lit(*r)),
        Expr::Bool(b) => Ok(if *b { "true".into() } else { "false".into() }),
        Expr::Str(s) => Ok(str_lit(s)),

        Expr::Not(inner) => Ok(format!("(not {})", expr(inner, env)?)),

        Expr::Binary(op, a, b) => binary(op, a, b, env),

        Expr::Ternary(c, t, f) => {
            let cs = expr(c, env)?;
            let ts = expr(t, env)?;
            let fs = expr(f, env)?;
            Ok(format!("(ite {cs} {ts} {fs})"))
        }

        // `lhs ∈ {a, b, c}` → (or (= lhs a) ...); `lhs ∈ {lo..hi}` → range bound.
        Expr::InExpr(lhs, rhs) => in_expr(lhs, rhs, env),

        // Everything below is out of the prototype subset.
        Expr::SetLit(_) => err("set literal (not as ∈ RHS) unsupported"),
        Expr::SeqLit(_) => err("sequence literal unsupported"),
        Expr::Range(..) => err("bare range unsupported (only as ∈ RHS)"),
        Expr::Tuple(_) => err("tuple unsupported"),
        Expr::Forall(..) => err("∀ quantifier unsupported (quantifier-free subset)"),
        Expr::Exists(..) => err("∃ quantifier unsupported (quantifier-free subset)"),
        Expr::Call(n, _) => err(format!("call `{n}` unsupported")),
        Expr::Cardinality(_) => err("cardinality `#` unsupported"),
        Expr::Index(..) => err("indexing `[]` unsupported"),
        Expr::Field(..) => err("field access unsupported (records out of subset)"),
        Expr::Match(..) => err("match unsupported"),
        Expr::Matches(..) => err("matches recognizer unsupported"),
        Expr::RunFsm { .. } => err("run(fsm) unsupported"),
    }
}

fn binary(op: &BinOp, a: &Expr, b: &Expr, env: &Env) -> Result<String, SmtLibError> {
    // `≠` has no SMT primitive; lower to (not (= ..)).
    if *op == BinOp::Neq {
        let (x, y) = (expr(a, env)?, expr(b, env)?);
        return Ok(format!("(not (= {x} {y}))"));
    }
    let sym = match op {
        BinOp::Eq => "=",
        BinOp::Lt => "<",
        BinOp::Le => "<=",
        BinOp::Gt => ">",
        BinOp::Ge => ">=",
        BinOp::And => "and",
        BinOp::Or => "or",
        BinOp::Implies => "=>",
        BinOp::Add => "+",
        BinOp::Sub => "-",
        BinOp::Mul => "*",
        // `++` string concatenation → Z3 seq concat.
        BinOp::Concat => "str.++",
        BinOp::Div => {
            // Int division is `div`; Real division is `/`. Infer from operand sorts.
            match sort_of(&Expr::Binary(BinOp::Add, Box::new(a.clone()), Box::new(b.clone())), env) {
                Some(Sort::Real) => "/",
                _ => "div",
            }
        }
        BinOp::Neq => unreachable!("handled above"),
    };
    let (x, y) = (expr(a, env)?, expr(b, env)?);
    Ok(format!("({sym} {x} {y})"))
}

fn in_expr(lhs: &Expr, rhs: &Expr, env: &Env) -> Result<String, SmtLibError> {
    let l = expr(lhs, env)?;
    match rhs {
        Expr::Range(lo, hi) => {
            let lo = expr(lo, env)?;
            let hi = expr(hi, env)?;
            Ok(format!("(and (>= {l} {lo}) (<= {l} {hi}))"))
        }
        Expr::SetLit(elems) => {
            if elems.is_empty() {
                return Ok("false".into());
            }
            let mut parts = Vec::with_capacity(elems.len());
            for el in elems {
                parts.push(format!("(= {l} {})", expr(el, env)?));
            }
            if parts.len() == 1 {
                Ok(parts.pop().unwrap())
            } else {
                Ok(format!("(or {})", parts.join(" ")))
            }
        }
        _ => err("∈ RHS must be a set literal or range in the prototype subset"),
    }
}

/// SMT-LIB int literal — negatives wrap in `(- n)`.
fn int_lit(i: i64) -> String {
    if i < 0 {
        // i64::MIN guard: format the magnitude as u64 to avoid overflow on negate.
        format!("(- {})", (i as i128).unsigned_abs())
    } else {
        i.to_string()
    }
}

/// SMT-LIB real literal — must carry a decimal point; negatives wrap in `(- r)`.
fn real_lit(r: f64) -> String {
    let mag = r.abs();
    let mut s = format!("{mag}");
    if !s.contains('.') && !s.contains('e') && !s.contains('E') {
        s.push_str(".0");
    }
    if r.is_sign_negative() && r != 0.0 {
        format!("(- {s})")
    } else {
        s
    }
}

/// SMT-LIB string literal: double-quoted, internal `"` doubled. Non-ASCII left
/// as-is (Z3 accepts UTF-8 in string literals); the prototype corpus is ASCII.
fn str_lit(s: &str) -> String {
    let mut out = String::with_capacity(s.len() + 2);
    out.push('"');
    for c in s.chars() {
        if c == '"' {
            out.push('"');
        }
        out.push(c);
    }
    out.push('"');
    out
}

// ---------------------------------------------------------------------------
// Solve: SMT-LIB text → Z3 → sat/unsat (+ model)
// ---------------------------------------------------------------------------

// `z3::Context` is a single-field newtype around `Z3_context`; we reach the raw
// pointer to query Z3's error state after parsing. The assert guards the layout.
const _: () = {
    assert!(
        std::mem::size_of::<Context>() == std::mem::size_of::<z3_sys::Z3_context>(),
        "z3::Context is no longer a single-pointer newtype; raw_ctx is unsound"
    );
};

#[inline]
fn raw_ctx(ctx: &Context) -> z3_sys::Z3_context {
    // SAFETY: layout verified by the const assert above.
    unsafe { *(ctx as *const Context as *const z3_sys::Z3_context) }
}

/// Run a claim through the full prototype path: emit SMT-LIB, parse it with Z3,
/// solve, and (when sat) extract the scalar model. Errors if the claim is out of
/// the transpilable subset, or if Z3 rejects the generated SMT-LIB.
pub fn solve(schema: &SchemaDecl) -> Result<SmtSolveResult, SmtLibError> {
    let (text, declared) = emit(schema)?;

    let mut cfg = Config::new();
    cfg.set_model_generation(true);
    // Serialized through the global setup lock (see crate::z3_ctx) so concurrent
    // creation never races Z3's global init.
    let ctx = { let _g = crate::z3_ctx::setup_guard(); Context::new(&cfg) };
    let solver = Solver::new(&ctx);

    solver.from_string(text.clone());

    // Detect a parser rejection — the z3 crate wrapper swallows it, so check the
    // raw error code. A malformed emit would otherwise silently add no assertions.
    let code = unsafe { z3_sys::Z3_get_error_code(raw_ctx(&ctx)) };
    if code != z3_sys::ErrorCode::OK {
        let msg = unsafe {
            let p = z3_sys::Z3_get_error_msg(raw_ctx(&ctx), code);
            std::ffi::CStr::from_ptr(p).to_string_lossy().into_owned()
        };
        return err(format!("Z3 rejected generated SMT-LIB ({code:?}): {msg}\n{text}"));
    }

    let satisfied = match solver.check() {
        SatResult::Sat => true,
        SatResult::Unsat => false,
        SatResult::Unknown => {
            return err("Z3 returned Unknown on a quantifier-free scalar problem");
        }
    };

    let mut bindings = HashMap::new();
    if satisfied {
        if let Some(model) = solver.get_model() {
            for (name, sort) in &declared {
                if let Some(v) = read_const(&ctx, &model, name, *sort) {
                    bindings.insert(name.clone(), v);
                }
            }
        }
    }

    Ok(SmtSolveResult { satisfied, bindings, smtlib: text })
}

/// Read one declared const out of the model by reconstructing its handle (Z3
/// consts are identified by name + sort, so a fresh `*::new_const` resolves to
/// the symbol the parser created).
fn read_const(ctx: &Context, model: &z3::Model, name: &str, sort: Sort) -> Option<Value> {
    use z3::ast::{Bool, Int, Real, String as Z3Str};
    match sort {
        Sort::Int => {
            let c = Int::new_const(ctx, name);
            model.eval(&c, true)?.as_i64().map(Value::Int)
        }
        Sort::Bool => {
            let c = Bool::new_const(ctx, name);
            model.eval(&c, true)?.as_bool().map(Value::Bool)
        }
        Sort::Real => {
            let c = Real::new_const(ctx, name);
            let (num, den) = model.eval(&c, true)?.as_real()?;
            Some(Value::Real(num as f64 / den as f64))
        }
        Sort::Str => {
            let c = Z3Str::new_const(ctx, name);
            model.eval(&c, true)?.as_string().map(Value::Str)
        }
    }
}
